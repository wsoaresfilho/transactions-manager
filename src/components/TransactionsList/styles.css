.transaction-list__list {
  list-style: none;
  padding: 5px;
  display: flex;
  flex-direction: column;
  height: 300px;
  overflow: auto;
}

.transaction-list__list::-webkit-scrollbar-track {
  -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
  background-color: var(--light-bg-color-secondary);
}

.dark .transaction-list__list::-webkit-scrollbar-track {
  background-color: var(--dark-bg-color-secondary);
}

.transaction-list__list::-webkit-scrollbar {
  width: 6px;
  background-color: #f5f5f5;
}

.transaction-list__list::-webkit-scrollbar-thumb {
  background-color: var(--header-bg-color);
}

.transaction-list__list-title {
  display: flex;
  align-items: center;
  padding: 5px 10px;
  min-height: 25px;
  background-color: #2b7d6a;
  color: ivory;
}

.transaction-list__list-title .transaction-list__item-badge {
  font-size: 1em;
  border: none;
  text-transform: capitalize;
}

.transaction-list__list-title .transaction-list__item-delete-btn {
  visibility: hidden;
}

.transaction-list_empty.transaction-list__list-item {
  text-align: center;
  display: block;
  padding: 40px;
}

.transaction-list__list-item {
  display: flex;
  align-items: center;
  padding: 5px 10px;
  min-height: 25px;
}

.transaction-list__list-item:nth-child(odd) {
  background: var(--light-bg-color);
}

.dark .transaction-list__list-item:nth-child(odd) {
  background: var(--dark-bg-color);
}

.transaction-list__item-description {
  width: 70%;
  white-space: nowrap;
  overflow: hidden;
  -o-text-overflow: ellipsis;
  text-overflow: ellipsis;
  padding-right: 15px;
}

.transaction-list__item-badge {
  text-transform: lowercase;
  font-size: 0.8em;
  border: 1px solid;
  width: 15%;
  min-width: 60px;
  text-align: center;
  padding: 2px;
}

.debit .transaction-list__item-badge {
  border-color: var(--debit-color);
  color: var(--debit-color);
}

.credit .transaction-list__item-badge {
  border-color: var(--credit-color);
  color: var(--credit-color);
}

.transaction-list__item-value {
  width: 15%;
  padding-left: 15px;
  text-align: right;
  min-width: 50px;
}

.transaction-list__item-delete-btn {
  background-color: var(--main-bg-color);
  padding: 2px 6px 2px 5px;
  font-weight: 600;
  text-align: right;
  margin-left: 15px;
  font-size: 12px;
  cursor: pointer;
  transition: all 0.3s ease-in-out;
  color: var(--debit-color);
  border: 1px solid red;
  border-radius: 20px;
  min-height: 20px;
  min-width: 20px;
}

.transaction-list__item-delete-btn:hover,
.transaction-list__item-delete-btn:focus {
  background-color: var(--debit-color);
  color: var(--light-bg-color);
}

.dark .transaction-list__item-delete-btn:hover,
.dark .transaction-list__item-delete-btn:focus {
  background-color: var(--debit-color);
  color: var(--dark-bg-color);
}

.transaction-list__footer {
  display: flex;
  justify-content: space-between;
  min-height: 50px;
  align-items: center;
  background: var(--header-bg-color);
  padding: 8px 25px;
  font-weight: bold;
  color: var(--light-bg-color-secondary);
}
