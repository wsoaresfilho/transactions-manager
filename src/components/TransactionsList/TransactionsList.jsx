import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import { moneyFormat, moneyFormatIntl } from '../../utils/helpers';
import './styles.css';

class TransactionsList extends PureComponent {
    constructor(props) {
        super(props);
        this.state = {};
    }

    handleDeleteClick(id) {
        const { deleteTransaction } = this.props;
        deleteTransaction(id);
    }

    render() {
        const { allTransactions, className } = this.props;
        const containerClassnames = `transaction-list__container ${className}`;
        const totalBalance = allTransactions.reduce((acc, curr) => {
            return acc + curr.value;
        }, 0);

        return (
            <div>
                <div className={containerClassnames}>
                    <ul className='transaction-list__list'>
                        <li
                            key='list-item-title'
                            className='transaction-list__list-title'
                        >
                            <span className='transaction-list__item-description'>
                                Description
                            </span>
                            <span className='transaction-list__item-badge'>
                                Type
                            </span>
                            <span className='transaction-list__item-value'>
                                Value
                            </span>
                            <button
                                className='transaction-list__item-delete-btn'
                                type='button'
                            >
                                X
                            </button>
                        </li>
                        {allTransactions.length === 0 && (
                            <span className='transaction-list_empty transaction-list__list-item'>
                                There are no item registered
                            </span>
                        )}
                        {allTransactions.reverse().map(item => {
                            const isCredit = item.value >= 0;
                            const itemClasses = classnames({
                                'transaction-list__list-item': true,
                                debit: !isCredit,
                                credit: isCredit,
                            });
                            const maneyValue = moneyFormat(item.value);

                            return (
                                <li key={item.id} className={itemClasses}>
                                    <span className='transaction-list__item-description'>
                                        {item.description}
                                    </span>
                                    <span className='transaction-list__item-badge'>
                                        {isCredit ? 'Credit' : 'Debit'}
                                    </span>
                                    <span className='transaction-list__item-value'>
                                        {maneyValue}
                                    </span>
                                    <span hidden className='hidden-id'>
                                        {item.id}
                                    </span>
                                    <button
                                        className='transaction-list__item-delete-btn'
                                        type='button'
                                        title='Delete transaction'
                                        onClick={() =>
                                            this.handleDeleteClick(item.id)
                                        }
                                    >
                                        X
                                    </button>
                                </li>
                            );
                        })}
                    </ul>
                </div>
                <div className='transaction-list__footer'>
                    <span>Final Balance:</span>
                    <span>{moneyFormatIntl(totalBalance)}</span>
                </div>
            </div>
        );
    }
}

TransactionsList.defaultProps = {
    className: null,
};

TransactionsList.propTypes = {
    allTransactions: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.arrayOf(
            PropTypes.shape({
                description: PropTypes.string,
                value: PropTypes.number,
                id: PropTypes.string,
            })
        ),
    ]).isRequired,
    deleteTransaction: PropTypes.func.isRequired,
    className: PropTypes.string,
};

export default TransactionsList;
