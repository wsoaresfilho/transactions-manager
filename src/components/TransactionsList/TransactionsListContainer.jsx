import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import TransactionsList from './TransactionsList';
import {
    deleteTransaction as deleteTransactionAction,
    fetchTransactionsData,
} from '../../actions';

export class TransactionsListContainer extends PureComponent {
    componentDidMount() {
        const { allTransactions, getAllTransactions } = this.props;
        if (allTransactions.length === 0) {
            getAllTransactions();
        }
    }

    render() {
        const { allTransactions, deleteTransaction, className } = this.props;
        return (
            <TransactionsList
                allTransactions={allTransactions}
                deleteTransaction={deleteTransaction}
                className={className}
            />
        );
    }
}

TransactionsListContainer.defaultProps = {
    className: null,
};

TransactionsListContainer.propTypes = {
    allTransactions: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.arrayOf(
            PropTypes.shape({
                description: PropTypes.string,
                value: PropTypes.number,
                id: PropTypes.string,
            })
        ),
    ]).isRequired,
    getAllTransactions: PropTypes.func.isRequired,
    deleteTransaction: PropTypes.func.isRequired,
    className: PropTypes.string,
};

const mapStateToProps = ({ transactions }) => {
    return {
        allTransactions: transactions.allTransactions,
    };
};

const mapDispatchToProps = dispatch => {
    return {
        getAllTransactions: () => dispatch(fetchTransactionsData()),
        deleteTransaction: transactionId =>
            dispatch(deleteTransactionAction(transactionId)),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(TransactionsListContainer);
