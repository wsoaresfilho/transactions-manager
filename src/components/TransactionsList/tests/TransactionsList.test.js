import React from 'react';
import { shallow } from 'enzyme';
import TransactionsList from '../TransactionsList';

describe('RegisterTransaction component', () => {
    let wrapper;
    let instance;
    let spy;
    const props = {
        deleteTransaction: jest.fn(),
    };

    const mockTransactions = [
        {
            description: 'item 1',
            value: 10,
            id: 'id1',
        },
        {
            description: 'item 2',
            value: -20,
            id: 'id2',
        },
        {
            description: 'item 3',
            value: 110.5,
            id: 'id3',
        },
    ];

    beforeEach(() => {
        wrapper = shallow(
            <TransactionsList
                deleteTransaction={props.deleteTransaction}
                allTransactions={[...mockTransactions]}
            />
        );
        instance = wrapper.instance();
        spy = {
            delete: jest.spyOn(TransactionsList.prototype, 'handleDeleteClick'),
        };
    });

    afterEach(() => {
        spy.delete.mockClear();
    });

    describe('when there are no transactions', () => {
        it('should render', () => {
            wrapper = shallow(
                <TransactionsList
                    deleteTransaction={props.deleteTransaction}
                    allTransactions={[]}
                />
            );

            expect(wrapper.find('.transaction-list__container').length).toEqual(
                1
            );
            expect(wrapper.find('.transaction-list__list').length).toEqual(1);
            expect(wrapper.find('.transaction-list_empty').text()).toEqual(
                'There are no items registered'
            );
            expect(wrapper.find('.transaction-list__list-title').length).toBe(
                1
            );
            expect(
                wrapper.find('.transaction-list__item-description').text()
            ).toEqual('Description');
            expect(
                wrapper.find('.transaction-list__item-badge').text()
            ).toEqual('Type');
            expect(
                wrapper.find('.transaction-list__item-value').text()
            ).toEqual('Value');
            // expect(
            //     wrapper.exists('.transaction-list__item-delete-btn')
            // ).toBeTruthy();
        });

        it('snapshot correctly', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there are some transactions', () => {
        it('should render', () => {
            expect(wrapper.exists('.transaction-list_empty')).toBeFalsy();
            expect(wrapper.find('.transaction-list__list-item').length).toEqual(
                mockTransactions.length
            );
            expect(
                wrapper
                    .find('.transaction-list__item-description')
                    .at(3)
                    .text()
            ).toEqual(mockTransactions[0].description);
            expect(
                wrapper
                    .find('.transaction-list__item-badge')
                    .at(3)
                    .text()
            ).toBe('Credit');
            expect(
                wrapper
                    .find('.transaction-list__item-value')
                    .at(3)
                    .text()
            ).toEqual('10.00');
            expect(
                wrapper
                    .find('.hidden-id')
                    .at(2)
                    .text()
            ).toEqual(mockTransactions[0].id);

            expect(
                wrapper
                    .find('.transaction-list__item-description')
                    .at(2)
                    .text()
            ).toEqual(mockTransactions[1].description);
            expect(
                wrapper
                    .find('.transaction-list__item-badge')
                    .at(2)
                    .text()
            ).toBe('Debit');
            expect(
                wrapper
                    .find('.transaction-list__item-value')
                    .at(2)
                    .text()
            ).toEqual('-20.00');
            expect(
                wrapper
                    .find('.hidden-id')
                    .at(1)
                    .text()
            ).toEqual(mockTransactions[1].id);

            expect(
                wrapper
                    .find('.transaction-list__item-description')
                    .at(1)
                    .text()
            ).toEqual(mockTransactions[2].description);
            expect(
                wrapper
                    .find('.transaction-list__item-badge')
                    .at(1)
                    .text()
            ).toBe('Credit');
            expect(
                wrapper
                    .find('.transaction-list__item-value')
                    .at(1)
                    .text()
            ).toEqual('110.50');
            expect(
                wrapper
                    .find('.hidden-id')
                    .at(0)
                    .text()
            ).toEqual(mockTransactions[2].id);

            expect(
                wrapper.find('.transaction-list__item-delete-btn').length
            ).toEqual(mockTransactions.length + 1);
        });

        describe('and click in the first delete button', () => {
            it('should delete the first transaction item', () => {
                wrapper
                    .find('.transaction-list__item-delete-btn')
                    .at(1)
                    .simulate('click');

                expect(spy.delete).toHaveBeenCalledTimes(1);
                expect(instance.props.deleteTransaction).toHaveBeenCalledWith(
                    mockTransactions[2].id
                );
            });
        });
    });
});
