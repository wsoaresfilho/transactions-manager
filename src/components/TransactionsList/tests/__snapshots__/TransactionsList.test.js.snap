// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterTransaction component when there are no transactions snapshot correctly 1`] = `ShallowWrapper {}`;
